### ISSUE TITLE
Passing a function to navigate with a call to setState doesn't actually run setState

### ISSUE THREAD

**Issue by ColinRyan** [react-navigation/react-navigation#2985]
I'm having an issue where I pass a function to navigate as a param and the function has a this.setState it in, along with a console.log. If I attach the function to the screen navigate is referencing and use it as part of an  onPress event on a text component, console.log will fire but the setState wont.

The function it's self will work however, if I use it as part of an onPress event on the component where the function is defined.

I'm very confused as to why this would be. I've seen issue https://github.com/react-community/react-navigation/issues/887
and my understand is that I'm doing things correctly (my code is nearly exactly the same as the code shown in a response on that issue as I've intentionally tried to mimic it to see if that might fix my issue.)

I'm using create-react-native-app with expo. Not sure if that makes a difference or not.

Thanks for your time!

### Your Environment

| software         | version
| ---------------- | -------
| react-navigation | ^1.0.0-beta.19
| react-native     | ^0.49.5
| node             | 8.8.1
| npm              | 5.5.1

**Comment by kelset** [comment]
1. can you share the code?
1. tbh it seems odd to me that you are able to send a setState method "across" components, my guts tell me that a setState will only work in the realm of the component it has been coded in, and not in the receiving end of it. (if this makes sense)
1. maybe it's also related to [setState being async](https://medium.com/@wereHamster/beware-react-setstate-is-asynchronous-ce87ef1a9cf3)?

**Comment by vonovak** [comment]
@ColinRyan what you're doing (passing setState of screen A to screen B) seems very inelegant, but then again, it should work. Are you sure that `this` in the function is the correct one?
please post a repro on snack, otherwise we cannot help

**Comment by ColinRyan** [comment]
I'm not passing setState, I'm passing a callback function with setState in it ;P

The repo is private but, I'll see about making a public repo to show off the issue. 

In the mean time, this what it looks like:

```
addItem(item) {
  this.setState({item})
  console.log(item)
}
//... further down in the render function

<Button onPress={() => navigate('Screen2', { 'addItem': (item) => this.addItem(item) }) }/>

//... in the Screen 2 component
const {params} = this.props.navigation.state;
//...
render() {
    const { params} = this.props.navigation.state;
    return (
         <View>
            <Text onPress={() => params.addItem('test')}> Some text </Text>
         </View>
    )
}

```

**Comment by luci1329** [comment]
I've just stumbled upon the same problem. If I navigate away from the component, it does not react to setState anymore (which is called in componentWillReceiveProps).

**Comment by thejacobseely** [comment]
@ColinRyan Let's see if I understood your question correctly. You want to pass a function as a prop to another screen, but still have the function modify the parent screen rather than the child screen, correct?

You need to bind the function to the screen you want it to modify when you pass it in.

For example, if you want addItem to modify the props on Screen 1 from Screen 2,  you need to pass it in as {"addItem": this.addItem.bind(this)}. This will bind the function to Screen 1, meaning that, when called, addItem will be run in the context of Screen 1 instead of Screen 2. Otherwise the program just runs addItem in the context of Screen 2 and setState will modify the state of the window from which it is run, rather than the state of Screen 1.

Hope that helps.

**Comment by luci1329** [comment]
@thejacobseely I don't think this is a matter of parent/child, screen 1 is not a parent of screen 2, they are both in the navigation stack.

**Comment by thejacobseely** [comment]
@luci1329 Fair, maybe my terminology was off, but either way in my experience that should solve the problem if I understand it correctly. Let me rephrase "You want to modify a screen on the stack from the screen which that screen opened,  correct?"

Binding the function when you pass it in as a parameter to the screen you navigate to should allow you to modify the original screen's state from the new screen.

**Comment by luci1329** [comment]
@thejacobseely Your solution would probably work for me but in the end it will just add more code when there is already a callback (aka componentWillReceiveProps) that should trigger a component re-rerender.

Let me explain my scenario. I have a redux connected component which is screen1 and there is screen2 which at some point gets pushed into the stack. Now if screen2 modifies something in the state(redux) to which screen1 is subscribed to, it is not re-rendered even though it received the new state through componentWillReceiveProps.

What I'm saying is that once a screen which is in the navigation stack, goes hidden, it will not re-render upon receiving new props.

**Comment by carsonwah** [comment]
I think you misunderstood some react lifecycle stuff. You are using `setState()` to modify the state, whereas props is a different thing in component. `props` are external which should not be changed by the component itself.

You may take a look at [official doc](https://reactjs.org/docs/react-component.html#componentwillreceiveprops). `componentWillReceiveProps()` is only invoked when component receive new props, not state. `shouldComponentUpdate()`, on the other hand, will always be triggered to check if it needs to re-render under new props or new state.

Also, remember to bind your function in constructor or use arrow function to reference `this`.

I have created a snack displaying a working situation of your case.   
https://snack.expo.io/S1_D_RWfG

**Comment by carsonwah** [comment]
I got confused with the code of @ColinRyan and @luci1329 . You are using redux, then it is not related to componentWillReceiveProps(). Maybe the binding issue really?

**Comment by luci1329** [comment]
@carsonwah You misunderstood what i wanted to say. No, I’m not callung setState in componentWillReceiveProps. I’m saying that my component is not rerendered with the new props if its not on the top of the navigation stack.

**Comment by carsonwah** [comment]
@luci1329 Yes, I got mixed up with you and @ColinRyan. Seems to be different case. For your case, I have written a minimal [snack](https://snack.expo.io/B17sTuMzf) with redux and react-navigation together, which seems working well. You may take a look.

Free feel to edit it to recreate your case.

**Comment by vonovak** [comment]
I'm closing this since this is not an issue with react-navigation and @carsonwah has provided a working solution for the OP's question.

**Comment by aashirMIQ** [comment]
having the same issue.
